### PR TITLE
split up the huge set of flags into smaller option structs

### PR DESCRIPTION
**Pull request by stewart-yu**
**What this PR does / why we need it**:
To make generic, we do following work:

1.  Spliting `KubeControllerManagerConfiguration` in kube-controller-manager and cloud-controller-manager into fewer smaller struct options order by controller, and modify relative flag. Also part of #59483.
2. Spliting `componentconfig` in controller-manager into fewer smaller config order by controller too.

All works follow #59582, using `option+config` logic.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cmd/controller-manager/app/config.go [hunk 1]
```diff
@@ -35,6 +36,8 @@ type Config struct {
 	Authentication  apiserver.AuthenticationInfo
 	Authorization   apiserver.AuthorizationInfo
 
+	CloudProvider *kubeoptions.CloudProviderOptions
```
Comment (deads2k) [inline cmd/controller-manager/app/config.go]: I think we should start creating the generic set and the set shared between these two binaries. Similar to how we created k8s.io/apiserver and k8s.io/kubernetes/pkg/kubeapiserver.

@sttts @cheftako

File: cmd/controller-manager/app/config.go [hunk 2]
```diff
@@ -46,6 +49,9 @@ type Config struct {
 
 	// the event sink
 	EventRecorder record.EventRecorder
+
+	EnableProfiling           bool
```
Comment (deads2k) [inline cmd/controller-manager/app/config.go]: Yes, these are generic.

File: cmd/controller-manager/app/options/feature_options.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+*/
+
+package options
+
+import (
+	"github.com/spf13/pflag"
+	genericcontrollermanager "k8s.io/kubernetes/cmd/controller-manager/app"
+)
+
+type FeatureOptions struct {
```
Comment (sttts) [inline cmd/controller-manager/app/options/feature_options.go]: these sound more like Debugging options. Features is a pretty neutral term. Moreover, we have something called "feature gates" in kube. So call this Feature might be confusing.

File: hack/.golint_failures [hunk 4]
```diff
@@ -1,4 +1,5 @@
 cluster/images/etcd-version-monitor
+cmd/controller-manager/app/options
```
Comment (sttts) [inline hack/.golint_failures]: can you fix the golint errors instead of adding it here?

File: cmd/controller-manager/app/options/feature_options.go (hunk 3 above)
Comment (stewart-yu) [reply cmd/controller-manager/app/options/feature_options.go]: done

File: hack/.golint_failures (hunk 4 above)
Comment (stewart-yu) [reply hack/.golint_failures]: fixed now

File: cmd/kube-controller-manager/app/options/options_test.go [hunk 5]
```diff
@@ -135,7 +135,7 @@ func TestAddFlags(t *testing.T) {
 				ConcurrentSATokenSyncs:                          10,
 				ConcurrentRCSyncs:                               10,
 				ConfigureCloudRoutes:                            false,
-				EnableContentionProfiling:                       true,
+				EnableContentionProfiling:                       false,
```
Comment (sttts) [inline cmd/kube-controller-manager/app/options/options_test.go]: is this change intentional?

File: cmd/controller-manager/app/options/options.go [hunk 6]
```diff
@@ -228,7 +234,7 @@ func (o *GenericControllerManagerOptions) Validate() []error {
 	errors = append(errors, o.InsecureServing.Validate()...)
 	errors = append(errors, o.Authentication.Validate()...)
 	errors = append(errors, o.Authorization.Validate()...)
-
+	errors = append(errors, o.Debugging.Validate()...)
```
Comment (sttts) [inline cmd/controller-manager/app/options/options.go]: nit: the order of calls should always be the same. Here Debugging is the last. For ApplyTo it is the first.

File: cmd/kube-controller-manager/app/options/options_test.go (hunk 5 above)
Comment (stewart-yu) [reply cmd/kube-controller-manager/app/options/options_test.go]: Yes. like `port` and `address`,  we should set it as value in `NewCloudControllerManagerOptions()`

**Comment by stewart-yu** [comment]
Have  some compile problem, just help me verify those split options struct right or not. thanks @sttts @deads2k

File: cmd/cloud-controller-manager/app/options/options.go [hunk 7]
```diff
@@ -36,7 +35,7 @@ import (
 
 // CloudControllerManagerOptions is the main context object for the controller manager.
 type CloudControllerManagerOptions struct {
-	Generic cmoptions.GenericControllerManagerOptions
+	GenericControllerManagerOptions *cmoptions.ControllerManagerOptions
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/options/options.go]: would keep the field short as `Generic`. The context here is clear.

File: cmd/controller-manager/app/options/options.go [hunk 8]
```diff
@@ -153,40 +171,46 @@ func NewDefaultControllerManagerComponentConfig(insecurePort int32) componentcon
...
-	fs.StringVar(&o.ComponentConfig.CloudConfigFile, "cloud-config", o.ComponentConfig.CloudConfigFile, "The path to the cloud provider configuration file. Empty string for no configuration file.")
-	fs.BoolVar(&o.ComponentConfig.AllowUntaggedCloud, "allow-untagged-cloud", false, "Allow the cluster to run without the cluster-id on cloud instances. This is a legacy mode of operation and a cluster-id will be required in the future.")
+func (o *ControllerManagerOptions) AddFlags(fs *pflag.FlagSet) {
+	fs.Int32Var(&o.ConcurrentServiceSyncs, "concurrent-service-syncs", o.ConcurrentServiceSyncs, "The number of services that are allowed to sync concurrently. Larger number = more responsive service management, but more CPU (and network) load")
+	// TODO: remove --service-account-private-key-file 6 months after 1.8 is released (~1.10)
+	fs.StringVar(&o.ServiceAccountKeyFile, "service-account-private-key-file", o.ServiceAccountKeyFile, "Filename containing a PEM-encoded private RSA or ECDSA key used to sign service account tokens.")
```
Comment (sttts) [inline cmd/controller-manager/app/options/options.go]: hadn't these been specific to the cloud controller-manager? Why are they now in the generic struct?

Comment (stewart-yu) [reply cmd/controller-manager/app/options/options.go]: Those two flag has exits in kube-controller-manager, so i remove those. sorry, i will revert it

File: cmd/kube-controller-manager/app/options/concurrentresourcessyncs_options.go [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+	ConcurrentResourceQuotaSyncs int32
+}
+
+// AddFlags adds flags related to debugging for controller manager to the specified FlagSet
+func (o *ConcurrentResourcesSyncsOptions) AddFlags(fs *pflag.FlagSet) {
+	if o == nil {
+		return
+	}
+
+	fs.Int32Var(&o.ConcurrentEndpointSyncs, "concurrent-endpoint-syncs", o.ConcurrentEndpointSyncs, "The number of endpoint syncing operations that will be done concurrently. Larger number = faster endpoint updating, but more CPU (and network) load")
```
Comment (sttts) [inline cmd/kube-controller-manager/app/options/concurrentresourcessyncs_options.go]: I tend to order them by controller, i.e. orthogonally to your order. @deads2k opinions?

Comment (deads2k) [reply cmd/kube-controller-manager/app/options/concurrentresourcessyncs_options.go]: > I tend to order them by controller, i.e. orthogonally to your order. @deads2k opinions?

Yes, order by controller so we can start looking for commonality and split them out by controller too.

**Comment by thockin** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/60270)

File: cmd/cloud-controller-manager/app/options/options.go (hunk 7 above)
Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/options/options.go]: revert it

File: cmd/kube-controller-manager/app/options/concurrentresourcessyncs_options.go (hunk 9 above)
Comment (stewart-yu) [reply cmd/kube-controller-manager/app/options/concurrentresourcessyncs_options.go]: done

File: cmd/cloud-controller-manager/app/options/options_test.go [hunk 10]
```diff
@@ -124,26 +157,25 @@ func TestAddFlags(t *testing.T) {
...
+			},
+			ResourceQuotaController: &cmoptions.ResourceQuotaControllerOptions{
+				ResourceQuotaSyncPeriod:      metav1.Duration{Duration: 5 * time.Minute},
+				ConcurrentResourceQuotaSyncs: 5,
+			},
+			SAController: &cmoptions.SAControllerOptions{
+				ConcurrentSATokenSyncs: 5,
+			},
+			Controllers:              []string{"*"},
+			ConcurrentDaemonSetSyncs: 2,
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/options/options_test.go]: are these 4 shared between controllers?

File: cmd/controller-manager/app/config.go [hunk 11]
```diff
@@ -17,17 +17,177 @@ limitations under the License.
...
+}
+
+// SAControllerInfo is part of context object for the controller manager.
+type SAControllerInfo struct {
+	ConcurrentSATokenSyncs int32
+}
+
+// ComponentConfigInfo is the main context object for the controller manager.
+type ComponentConfigInfo struct {
+	GenericComponentConfig                 GenericComponentInfo
```
Comment (sttts) [inline cmd/controller-manager/app/config.go]: do we need the Config postfix in those fields?

File: cmd/controller-manager/app/config.go [hunk 12]
```diff
@@ -17,17 +17,177 @@ limitations under the License.
...
+	ServiceCIDR                            string
+	NodeCIDRMaskSize                       int32
+	RootCAFile                             string
+}
+
 // Config is the main context object for the controller manager.
 type Config struct {
-	// TODO: split up the component config. This is not generic.
-	ComponentConfig componentconfig.KubeControllerManagerConfiguration
+	ComponentConfig ComponentConfigInfo
```
Comment (sttts) [inline cmd/controller-manager/app/config.go]: so we got rid of the `KubeControllerManagerConfiguration` configuration? I think we have to move this great new structure to the api object.

For the controller manager we can change the component config API object easily because we have not persisted it anywhere. For the scheduler we already did that (as configuration file on disk).  So for the scheduler we have to create a new version and potentially convert it.

But he we should at least follow the component config API pattern.

File: cmd/cloud-controller-manager/app/options/options_test.go (hunk 10 above)
Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/options/options_test.go]: No, just in kube-controller-manager.

File: cmd/controller-manager/app/config.go (hunk 11 above)
Comment (stewart-yu) [reply cmd/controller-manager/app/config.go]: Now that we remove those struct into component api, delete postfix `Config` now

File: cmd/controller-manager/app/options/options.go [hunk 13]
```diff
@@ -195,63 +195,93 @@ func NewGenericControllerManagerOptions(componentConfig componentconfig.KubeCont
...
-				IncrementTimeoutHostPath: 30,
+		GenericComponent: componentconfig.GenericComponentConfiguration{
+			Port:                      insecurePort,
+			Address:                   "0.0.0.0",
+			RouteReconciliationPeriod: metav1.Duration{Duration: 10 * time.Second},
+			MinResyncPeriod:           metav1.Duration{Duration: 12 * time.Hour},
+			NodeMonitorPeriod:         metav1.Duration{Duration: 5 * time.Second},
+			ClusterName:               "kubernetes",
+			ConfigureCloudRoutes:      true,
+			ContentType:               "application/vnd.kubernetes.protobuf",
```
Comment (sttts) [inline cmd/controller-manager/app/options/options.go]: where is this used?

File: pkg/apis/componentconfig/types.go [hunk 14]
```diff
@@ -163,53 +163,62 @@ type GroupResource struct {
 type KubeControllerManagerConfiguration struct {
```
Comment (sttts) [inline pkg/apis/componentconfig/types.go]: do we have a version type in k8s.io/api as well?

Note: we want to move these types into cmd/ as a follow-up.

File: cmd/controller-manager/app/options/options.go (hunk 13 above)
Comment (stewart-yu) [reply cmd/controller-manager/app/options/options.go]: Struct `GenericComponentConfigOptions`  aims to put  some parameter both kube-controller-manager and cloud -controller-manager used. For example, `NodeMonitorPeriod` used in cloud-node-controller and nodelifeController, i think is difficulty to order it to specially  controller,  so just put it in `GenericComponentConfigOptions`. WDYT?

File: pkg/apis/componentconfig/types.go (hunk 14 above)
Comment (stewart-yu) [reply pkg/apis/componentconfig/types.go]: No, will mark it in follow-up later.

File: cmd/controller-manager/app/options/options.go [hunk 15]
```diff
@@ -91,91 +195,105 @@ func NewGenericControllerManagerOptions(componentConfig componentconfig.KubeCont
...
+			KubeAPIQPS:                20.0,
+			KubeAPIBurst:              30,
+			ControllerStartInterval:   metav1.Duration{Duration: 0 * time.Second},
+			ConcurrentServiceSyncs:    1,
```
Comment (sttts) [inline cmd/controller-manager/app/options/options.go]: is this used in both binaries?

File: cmd/controller-manager/app/options/options.go (hunk 13 above)
Comment (sttts) [reply cmd/controller-manager/app/options/options.go]: Have you checked all of these to be used in both controller managers?

I would prefer to have clearly generic values only here in the struct. E.g. MinResyncPeriod, ContentType, KubeAPIQPS, KubeAPIBurst, ControllerStartInterval, LeaderElection look generic. Think of we create another controller-manager with different controllers. Which values will it need as well?

All the others should probably be duplicated in both binaries.

File: cmd/controller-manager/app/options/options.go (hunk 15 above)
Comment (stewart-yu) [reply cmd/controller-manager/app/options/options.go]: Yes. sorry for reply delay.

File: cmd/controller-manager/app/options/options.go (hunk 13 above)
Comment (stewart-yu) [reply cmd/controller-manager/app/options/options.go]: emmm, little confused about it. Both controller-manager used it. I think it should better construct more well struct to split some parameter out `GenericComponentConfigOptions` to make `GenericComponentConfigOptions` more Generic.

File: cmd/controller-manager/app/options/options.go (hunk 15 above)
Comment (stewart-yu) [reply cmd/controller-manager/app/options/options.go]: I want to construct `serviceControllerOption` struct to hold  `ConcurrentServiceSyncs` parameter later.

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 16]
```diff
@@ -86,13 +86,13 @@ the cloud specific control loops shipped with Kubernetes.`,
...
 		factor := rand.Float64() + 1
-		return time.Duration(float64(c.Generic.ComponentConfig.MinResyncPeriod.Nanoseconds()) * factor)
+		return time.Duration(float64(c.Generic.ComponentConfig.GenericComponent.MinResyncPeriod.Nanoseconds()) * factor)
 	}
 }
 
 // Run runs the ExternalCMServer.  This should never exit.
 func Run(c *cloudcontrollerconfig.CompletedConfig) error {
-	cloud, err := cloudprovider.InitCloudProvider(c.Generic.ComponentConfig.CloudProvider, c.Generic.ComponentConfig.CloudConfigFile)
+	cloud, err := cloudprovider.InitCloudProvider(c.Generic.ComponentConfig.CloudProvider.CloudProvider, c.Generic.ComponentConfig.CloudProvider.CloudConfigFile)
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/controllermanager.go]: should `CloudProvider.CloudProvider` be better `CloudProvider.Name` ?

File: cmd/controller-manager/app/options/options.go [hunk 17]
```diff
@@ -42,8 +42,35 @@ import (
 // GenericControllerManagerOptions is the common structure for a controller manager. It works with NewGenericControllerManagerOptions
 // and AddDefaultControllerFlags to create the common components of kube-controller-manager and cloud-controller-manager.
 type GenericControllerManagerOptions struct {
-	// TODO: turn ComponentConfig into modular option structs. This is not generic.
-	ComponentConfig componentconfig.KubeControllerManagerConfiguration
+	GenericComponent                 *GenericComponentConfigOptions
```
Comment (sttts) [inline cmd/controller-manager/app/options/options.go]: Can we call this `Generic`?

File: cmd/controller-manager/app/options/options.go (hunk 13 above)
Comment (sttts) [reply cmd/controller-manager/app/options/options.go]: There is a difference between "generic" and "shared" options. The first are semantically generic options. The later are those that are shared because we would duplicate implementation otherwise.

Imagine you have a 3rdparty-controller-manager, managing external 3rdparty resources, e.g. a PaaS or Istio. Which options would they expect to have as well ("generic")? Which will be kube specific ("shared" between kube and cloud ctrl-manager) ?

File: cmd/controller-manager/app/options/options.go (hunk 17 above)
Comment (stewart-yu) [reply cmd/controller-manager/app/options/options.go]: Avoiding cite like`Generic.Generic`, because `CloudControllerManagerOptions` struct have `Generic` parameter. keep seems more better?

File: cmd/controller-manager/app/config.go (hunk 1 above)
Comment (stewart-yu) [reply cmd/controller-manager/app/config.go]: done, PTAL

File: cmd/cloud-controller-manager/app/controllermanager.go (hunk 16 above)
Comment (stewart-yu) [reply cmd/cloud-controller-manager/app/controllermanager.go]: fix now

File: cmd/controller-manager/app/options/options.go (hunk 13 above)
Comment (stewart-yu) [reply cmd/controller-manager/app/options/options.go]: thank for explanations, split it now

File: cmd/cloud-controller-manager/app/options/options_test.go [hunk 18]
```diff
@@ -63,55 +65,90 @@ func TestAddFlags(t *testing.T) {
...
-				ConcurrentSATokenSyncs:                          5,
-				ConcurrentServiceSyncs:                          1,
-				ConcurrentGCSyncs:                               20,
-				ConcurrentRCSyncs:                               5,
-				MinResyncPeriod:                                 metav1.Duration{Duration: 100 * time.Minute},
-				NodeMonitorPeriod:                               metav1.Duration{Duration: 5 * time.Second},
-				ResourceQuotaSyncPeriod:                         metav1.Duration{Duration: 5 * time.Minute},
-				NamespaceSyncPeriod:                             metav1.Duration{Duration: 5 * time.Minute},
-				PVClaimBinderSyncPeriod:                         metav1.Duration{Duration: 15 * time.Second},
+		Generic: &cmoptions.GenericControllerManagerOptions{
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/options/options_test.go]: Which of the field in `GenericControllerManagerOptions` are actually used in kube- **and** cloud-controller-manager. In other words, `GenericControllerManagerOptions` does not really look generic.

If we could get rid of the `GenericControllerManagerOptions` layer now, we would also get rid of the strange `Generic.GenericComponent`.

File: pkg/apis/componentconfig/types.go [hunk 19]
```diff
@@ -219,116 +234,144 @@ type KubeControllerManagerConfiguration struct {
 	// allowed to sync concurrently. Larger number = more responsive jobs,
 	// but more CPU (and network) load.
 	ConcurrentJobSyncs int32
```
Comment (sttts) [inline pkg/apis/componentconfig/types.go]: this needs a JobControllerConfiguration

File: pkg/apis/componentconfig/types.go [hunk 20]
```diff
@@ -219,116 +234,144 @@ type KubeControllerManagerConfiguration struct {
...
-	ConcurrentNamespaceSyncs int32
-	// concurrentSATokenSyncs is the number of service account token syncing operations
-	// that will be done concurrently.
-	ConcurrentSATokenSyncs int32
-	// lookupCacheSizeForRC is the size of lookup cache for replication controllers.
-	// Larger number = more responsive replica management, but more MEM load.
 	// nodeSyncPeriod is the period for syncing nodes from cloudprovider. Longer
 	// periods will result in fewer calls to cloud provider, but may delay addition
 	// of new nodes to cluster.
 	NodeSyncPeriod metav1.Duration
```
Comment (sttts) [inline pkg/apis/componentconfig/types.go]: this should be part of the KubeCloudShared, as it is used by multiple kube- and cloud-controllers.

File: pkg/apis/componentconfig/types.go (hunk 19 above)
Comment (stewart-yu) [reply pkg/apis/componentconfig/types.go]: Done. Should we also build configuration for ConcurrentDaemonSetSyncs?

File: pkg/apis/componentconfig/types.go (hunk 20 above)
Comment (stewart-yu) [reply pkg/apis/componentconfig/types.go]: Done. Now that it Mark as Deprecated, we can move it to KubeCloudShared.

File: pkg/apis/componentconfig/types.go [hunk 21]
```diff
@@ -164,171 +164,211 @@ type GroupResource struct {
...
-	// management, but more CPU (and network) load.
-	ConcurrentResourceQuotaSyncs int32
-	// concurrentDeploymentSyncs is the number of deployment objects that are
-	// allowed to sync concurrently. Larger number = more responsive deployments,
-	// but more CPU (and network) load.
-	ConcurrentDeploymentSyncs int32
 	// concurrentDaemonSetSyncs is the number of daemonset objects that are
 	// allowed to sync concurrently. Larger number = more responsive daemonset,
 	// but more CPU (and network) load.
 	ConcurrentDaemonSetSyncs int32
```
Comment (sttts) [inline pkg/apis/componentconfig/types.go]: this belong to a DaemonSetControllerConfiguration, doesn't it?

Comment (sttts) [reply pkg/apis/componentconfig/types.go]: @stewart-yu can you double check all remaining fields here whether they belong to some controller?

File: pkg/apis/componentconfig/types.go [hunk 22]
```diff
@@ -354,6 +343,25 @@ type DeploymentControllerConfiguration struct {
...
 }
 
+type DaemonSetControllerConfiguration struct {
+	// concurrentDaemonSetSyncs is the number of daemonset objects that are
+	// allowed to sync concurrently. Larger number = more responsive daemonset,
+	// but more CPU (and network) load.
+	ConcurrentDaemonSetSyncs int32
+}
+
+type DeprecatedComponentConfiguration struct {
```
Comment (sttts) [inline pkg/apis/componentconfig/types.go]: DeprecatedControllerConfiguration

File: pkg/apis/componentconfig/v1alpha1/types.go [hunk 23]
```diff
@@ -270,26 +279,6 @@ type KubeControllerManagerConfiguration struct {
 	// externalCloudVolumePlugin specifies the plugin to use when cloudProvider is "external".
```
Comment (sttts) [inline pkg/apis/componentconfig/v1alpha1/types.go]: much better now :+1:

File: cmd/controller-manager/app/options/options.go [hunk 24]
```diff
@@ -40,8 +41,32 @@ import (
...
+	JobController                    *JobControllerOptions
+	NamespaceController              *NamespaceControllerOptions
+	NodeIpamController               *NodeIpamControllerOptions
+	NodeLifecycleController          *NodeLifecycleControllerOptions
+	PersistentVolumeBinderController *PersistentVolumeBinderControllerOptions
+	ReplicaSet                       *ReplicaSetOptions
+	ReplicationController            *ReplicationControllerOptions
+	ResourceQuotaController          *ResourceQuotaControllerOptions
+	SAController                     *SAControllerOptions
+	metav1.TypeMeta
```
Comment (sttts) [inline cmd/controller-manager/app/options/options.go]: no need for the TypeMeta here. This is only for API types.

File: cmd/controller-manager/app/options/options.go [hunk 25]
```diff
@@ -40,8 +41,32 @@ import (
...
+	GenericComponent                 *GenericComponentConfigOptions
+	KubeCloudShared                  *KubeCloudSharedOptions
+	Debugging                        *DebuggingOptions
+	CloudProvider                    *CloudProviderOptions
+	ServiceController                *ServiceControllerOptions
+	AttachDetachController           *AttachDetachControllerOptions
+	CSRSigningController             *CSRSigningControllerOptions
+	DeploymentController             *DeploymentControllerOptions
+	DaemonSetController              *DaemonSetControllerOptions
+	DeprecatedComponent              *DeprecatedComponentOptions
```
Comment (sttts) [inline cmd/controller-manager/app/options/options.go]: DeprecatedFlags is better maybe.

File: cmd/controller-manager/app/options/options.go [hunk 26]
```diff
@@ -132,12 +232,81 @@ func (o *GenericControllerManagerOptions) AddFlags(fs *pflag.FlagSet) {
 
 // ApplyTo fills up controller manager config with options and userAgent
 func (o *GenericControllerManagerOptions) ApplyTo(c *genericcontrollermanager.Config, userAgent string) error {
-	c.ComponentConfig = o.ComponentConfig
-
+	c.ComponentConfig.Kind = o.Kind
+	c.ComponentConfig.APIVersion = o.APIVersion
```
Comment (sttts) [inline cmd/controller-manager/app/options/options.go]: no need for these two

File: cmd/controller-manager/app/options/replicaset_options.go [hunk 27]
```diff
@@ -0,0 +1,57 @@
+/*
```
Comment (sttts) [inline cmd/controller-manager/app/options/replicaset_options.go]: how about removing the _options suffix from all options go files here? As the package is called options, the suffix is redundant.

File: pkg/apis/componentconfig/types.go [hunk 28]
```diff
@@ -164,171 +164,219 @@ type GroupResource struct {
...
+	// KubeCloudSharedConfiguration holds configuration for shared related features
+	// both in cloud controller manager and kube-controller manager.
+	KubeCloudShared KubeCloudSharedConfiguration
+	// DebuggingConfiguration holds configuration for Debugging related features.
+	Debugging DebuggingConfiguration
+	// CloudProviderConfiguration holds configuration for CloudProvider related features.
+	CloudProvider CloudProviderConfiguration
+	// ServiceControllerConfiguration holds configuration for ServiceController
+	// related features.
+	ServiceController ServiceControllerConfiguration
```
Comment (sttts) [inline pkg/apis/componentconfig/types.go]: let's put an empty line above this first controllers. And then sort the controllers by name, same for the ControllerConfigurations below.

File: cmd/controller-manager/app/options/replicaset_options.go (hunk 27 above)
Comment (stewart-yu) [reply cmd/controller-manager/app/options/replicaset_options.go]: great! will remove, now that in `option` directory.

File: cmd/controller-manager/app/options/attachdetachcontroller_options.go [hunk 29]
```diff
@@ -0,0 +1,61 @@
...
+
+package options
+
+import (
+	"github.com/spf13/pflag"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/apis/componentconfig"
+)
+
+// AttachDetachControllerOptions is part of context object for the controller manager.
```
[the remark this answers is not in the document]
Comment (stewart-yu) [reply cmd/controller-manager/app/options/attachdetachcontroller_options.go]: how about `AttachDetachControllerOptions holds the AttachDetachController options`?

File: pkg/apis/componentconfig/types.go (hunk 19 above)
Comment (sttts) [reply pkg/apis/componentconfig/types.go]: it's part of the DaemonSetController.

File: cmd/controller-manager/app/options/attachdetachcontroller_options.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (sttts) [reply cmd/controller-manager/app/options/attachdetachcontroller_options.go]: yes. not much info in there, but this sounds ok.

**Comment by sttts** [comment]
This is great now as a first step, a huge step forward. We will have a number of follow-ups, especially to
- adapt the options+componentconfig pattern to @mtaufen's proposal
- split the generic component config and options into a kube and cloud part.

/lgtm
/approve

/assign @deads2k 
/assign @luxas 
/assign @mtaufen

for approval.

**Comment by deads2k** [comment]
I think there is more to do, but this is a huge step forward and necessary to take stock of what we have before we split it further.

/approve
@cheftako you had some wiring pulls. I think this is logical next step when trying to find units of commonality as you started back in December.

**Review by luxas — APPROVED** [review]
Didn't have time to review this line by line, but I approve of the general direction here. After this PR is merged, even more optimization could be done, to better separate the parts from a "real" generic controller-manager, the k8s core controller-manager, and the cloud controller-manager

**Comment by stewart-yu** [comment]
Commit log rebased, thanks all. kindly ping @mikedanese approve for `pkg/apis`
